File: data-exercise/spark/spark-sql/src/test/java/org/mqjd/spark/java/spark01/HelloSparkDFTest.java
package org.mqjd.spark.java.spark01;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.junit.Test;
import org.mqjd.spark.java.SparkJobTest;

public class HelloSparkDFTest extends SparkJobTest {

    @Test
    public void given_expect_dataset_when_HelloSparkDF_then_success() {
        String targetDir = "target/spark01";
        HelloSparkDF.main(new String[]{"--output", targetDir});
        SparkSession spark = getSparkSession();
        Dataset<Row> expectResult = spark.read().option("delimiter", ",").option("header", "true")
            .csv(targetDir).toDF();
        Dataset<Row> actualResult = spark.read().option("delimiter", ",").option("header", "true")
            .csv(getResourceFile("/spark01.csv")).toDF();
        assertDatasetEquals(expectResult, actualResult);
    }

    @Test
    public void test() {
        SparkSession spark = getSparkSession();
        Dataset<Row> actualResult = spark.read().option("delimiter", ",").option("header", "true")
            .csv(getResourceFile("/spark01.csv")).toDF();
        Dataset<Row> df = actualResult.select(new Column("number").cast("long"));
        df = df.select(new Column("number").plus(2L).multiply(10).divide(3));
        df.printSchema();
        df.explain(true);
    }

}

File: data-exercise/spark/spark-sql/src/test/scala/org/mqjd/spark/scala/spark01/HelloSparkDFTest.scala
package org.mqjd.spark.scala.spark01

import org.apache.spark.sql.Row
import org.apache.spark.sql.functions.lit
import org.apache.spark.sql.types.{DoubleType, IntegerType, StringType, StructField, StructType}
import org.mqjd.spark.scala.SparkTest
import org.scalatest.funsuite.AnyFunSuite

class HelloSparkDFTest extends AnyFunSuite with SparkTest {
  test("xxx") {
    // Define schema
    val schema = StructType(Array(
      StructField("id", IntegerType, nullable = true),
      StructField("name", StringType, nullable = true),
      StructField("mount", DoubleType, nullable = true)
    ))
    // Define row data
    val data = Seq(
      Row(1, null, 622543.52127),
      Row(2, " \t asd   ", 622872.91534)
    )
    spark.createDataFrame(spark.sparkContext.parallelize(data), schema).createOrReplaceTempView("test")
    spark.sql("select regexp_replace(name, '[\\s\\t]+', '' ) from test").withColumn("a", lit("asdad")).show()
    println(" \t asd   ")
  }
  test("given expect dataset when HelloSparkDF then success") {
    val targetDir = getTargetFile("spark01")
    deleteDirectory(targetDir)
    val args = Array("--output", targetDir)
    HelloSparkDF.main(args)
    val expectResult =
      spark.read.option("delimiter", ",").option("header", "true").csv(targetDir).toDF()
    val actualResult =
      spark.read.option("delimiter", ",").option("header", "true").csv(getResourceFile("/spark01.csv")).toDF()
    assertDataFrameEquals(expectResult, actualResult)
  }
}
